src/osd/scheduler/OpSchedulerItem.cc

```diff
@@ -195,6 +195,12 @@ void PGRecovery::run(
   pg->unlock();
 }
 
+PGRecovery::~PGRecovery()
+{
+  if (reserved_pushes > 0)
+    osd->release_reserved_pushes(reserved_pushes);
```

Comment:
OSDService::release_reserved_pushes() takes a lock and invokes _maybe_queue_recovery().  I'm not sure I'm happy with doing that in a destructor.  OSDService::recovery_lock, however, is a leaf lock (never held while taking any other lock) and should be safe to take in basically all contexts.  Additionally, this PR is a significant simplification.  I need to look at this in a bit more detail. @rzarzynski What do you think?